client/java/src/main/java/io/openlineage/client/transports/TokenProviderTypeIdResolver.java

```diff
@@ -0,0 +1,50 @@
+/*
+/* Copyright 2018-2024 contributors to the OpenLineage project
+/* SPDX-License-Identifier: Apache-2.0
+*/
+
+package io.openlineage.client.transports;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.databind.DatabindContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.jsontype.impl.TypeIdResolverBase;
+import java.util.Objects;
+
+public class TokenProviderTypeIdResolver extends TypeIdResolverBase {
+  private JavaType superType;
+
+  @Override
+  public void init(JavaType baseType) {
+    superType = baseType;
+  }
+
+  @Override
+  public JsonTypeInfo.Id getMechanism() {
+    return JsonTypeInfo.Id.NAME;
+  }
+
+  @Override
+  public String idFromValue(Object obj) {
+    return idFromValueAndType(obj, obj.getClass());
+  }
+
+  @Override
+  public String idFromValueAndType(Object obj, Class<?> subType) {
+    return subType.getCanonicalName();
+  }
+
+  @Override
+  public JavaType typeFromId(DatabindContext context, String id) {
+    Class<?> subType = null;
+    if (Objects.equals(id, "api_key")) // backwards compatibility
+    return context.constructSpecializedType(superType, ApiKeyTokenProvider.class);
```

Comment:
don't forget our old friend  `Spotless`

Reply:
funny thing is, when you forget to put curly brackets after `if`, spotless doesn't even recognise it as incorrect and spotlessApply changes indentation to what you can see here